File: app/src/main/java/org/evcs/android/features/profile/wallet/AddCreditCardPresenter.java
package org.evcs.android.features.profile.wallet;

import com.base.networking.retrofit.RetrofitServices;

import org.evcs.android.network.service.presenter.ServicesPresenter;

public class AddCreditCardPresenter<T extends IBrainTreeView> extends ServicesPresenter<T> {

    private static final String TAG = "AddCreditCardPresenter";
    private static final String TOKEN_KEY = "token";

    protected static final String LOCATION_KEY = "location";

    private String mTokenUrl;

    public AddCreditCardPresenter(T brainTreeFragment, RetrofitServices services) {
        super(brainTreeFragment, services);
    }

    /**
     * Requests from API the url where the token will be found.
     */
    public void getTokenEndpoint() {
//        getService(BrainTreeService.class).getTokenEndpoint().enqueue(new AuthCallback<Void>(this) {
//            @Override
//            public void onResponse(Call<Void> call, Response<Void> response) {
//                super.onResponse(call, response);
//                Headers headers = response.headers();
//                mTokenUrl = headers.get(LOCATION_KEY);
//                if (isViewCreated()) {
//                    if (mTokenUrl == null) {
//                        getView().showError(ErrorUtils.getError(response.errorBody()));
//                    } else {
//                        getToken();
//                    }
//                }
//            }
//
//            @Override
//            public void onResponseSuccessful(Void aVoid) {
//                // This method must be empty.
//            }
//
//            @Override
//            public void onResponseFailed(ResponseBody responseBody, int i) {
//                getView().showError(ErrorUtils.getError(responseBody));
//            }
//
//            @Override
//            public void onCallFailure(Throwable throwable) {
//                Log.e(TAG, throwable.getMessage(), throwable);
//                runIfViewCreated(new Runnable() {
//                    @Override
//                    public void run() {
//                        getView().showError(RequestError.getNetworkError());
//                    }
//                });
//            }
//        });
    }

    /**
     * Queries the url to check for the BrainTree token. If it is not available, we will receive a
     * string instead of a JSON, which is why I have to use Object and do a weird casting. If it was
     * not available, we have to retry.
     */
    protected void getToken() {
//        new PollingManager(this).poll(getService(BrainTreeService.class).getToken(mTokenUrl), new PollingManager.PollingCallback() {
//            @Override
//            public void onResponseSuccessful(Response response) {
//                getView().onTokenResponse((String) ((LinkedTreeMap) response.body()).get(TOKEN_KEY));
//            }
//
//            @Override
//            public void onResponseFailed(ResponseBody responseBody, int i) {
//                getView().onTokenError();
//            }
//
//            @Override
//            public void onCallFailure() {
//                runIfViewCreated(new Runnable() {
//                    @Override
//                    public void run() {
//                        getView().onTokenError();
//                    }
//                });
//            }
//
//            @Override
//            public boolean shouldRetry(Response response) {
//                return response.code() == NetworkCodes.ACCEPTED;
//            }
//        });
    }

    public void getTokenAndMakeDefaultPaymentMethod(final String item) {
//        getService(BrainTreeService.class).getPaymentMethods()
//                .enqueue(new AuthCallback<List<CreditCard>>(this) {
//                    @Override
//                    public void onResponseSuccessful(List<CreditCard> creditCardInformations) {
//                        Collections.reverse(creditCardInformations);
//                        for (CreditCard cci : creditCardInformations) {
//                            if (cci.last4.equals(item.substring(item.length() - 4))) {
//                                makeDefaultPaymentMethod(cci.getToken());
//                                return;
//                            }
//                        }
//                    }
//
//                    @Override
//                    public void onResponseFailed(ResponseBody responseBody, int i) {
//                        getView().showError(ErrorUtils.getError(responseBody));
//                    }
//
//                    @Override
//                    public void onCallFailure(Throwable throwable) {
//                        getView().showError(RequestError.getNetworkError());
//                    }
//                });
    }

    public void makeDefaultPaymentMethod(final String item) {
//        getService(BrainTreeService.class).makeDefaultPaymentMethod(item)
//                .enqueue(new AuthCallback<List<CreditCard>>(this) {
//                    @Override
//                    public void onResponseSuccessful(List<CreditCard> creditCardInformations) {
//                        getView().onMakeDefaultFinished();
//                    }
//
//                    @Override
//                    public void onResponseFailed(ResponseBody responseBody, int i) {
//                        getView().showError(ErrorUtils.getError(responseBody));
//                    }
//
//                    @Override
//                    public void onCallFailure(Throwable throwable) {
//                        getView().showError(RequestError.getNetworkError());
//                    }
//                });
    }
}

File: app/src/main/java/org/evcs/android/features/profile/wallet/AddCreditCardPresenter.kt
package org.evcs.android.features.profile.wallet

import com.base.networking.retrofit.RetrofitServices
import com.stripe.android.model.Address
import com.stripe.android.model.ConfirmSetupIntentParams
import com.stripe.android.model.PaymentMethod
import com.stripe.android.model.PaymentMethodCreateParams
import okhttp3.ResponseBody
import org.evcs.android.Configuration
import org.evcs.android.model.ClientSecret
import org.evcs.android.network.service.presenter.ServicesPresenter
import org.evcs.android.network.service.PaymentMethodsService
import org.evcs.android.util.UserUtils
import org.evcs.android.network.callback.AuthCallback
import org.evcs.android.model.shared.RequestError
import org.evcs.android.util.ErrorUtils

class AddCreditCardPresenter(viewInstance: AddCreditCardView, services: RetrofitServices?)
    : CreditCardPresenter<AddCreditCardView>(viewInstance, services) {

    private var mClientSecret: String? = null

    fun getClientSecret() {
        getService(PaymentMethodsService::class.java).getClientSecret()
            .enqueue(object : AuthCallback<ClientSecret>(this) {
                override fun onResponseSuccessful(response: ClientSecret) {
                    mClientSecret = response.clientSecret
                }

                override fun onResponseFailed(responseBody: ResponseBody, code: Int) {
                    view.showError(ErrorUtils.getError(responseBody))
                }

                override fun onCallFailure(t: Throwable) {
                    view.showError(RequestError.getNetworkError())
                }
            })
        }

    fun getConfirmParams(name: String, card: PaymentMethodCreateParams.Card, zipcode: String): ConfirmSetupIntentParams {
        val billingDetails = PaymentMethod.BillingDetails.Builder()
            .setName(name)
//            .setEmail(UserUtils.getUserEmail())
            .setAddress(Address.Builder().setPostalCode(zipcode).build())
            .build()

        val paymentMethodParams = PaymentMethodCreateParams
            .create(card, billingDetails, null)

        return ConfirmSetupIntentParams
            .create(paymentMethodParams, mClientSecret!!)
    }
}